Initialize project

KyuubiRan opened the pull request.

KyuubiRan committed:
Initialize project

KyuubiRan committed:
feat: Add MethodFinder.java

KyuubiRan committed:
feat: Add ConstructorFinder.java

KyuubiRan committed:
refactor: change namespace

KyuubiRan committed:
refactor: change namespace

KyuubiRan committed:
refactor: remove old xposed library

yujincheng08 commented on gradle.properties:
```diff
@@ -0,0 +1,21 @@
+# http://www.gradle.org/docs/current/userguide/build_environment.html
+# Specifies the JVM arguments used for the daemon process.
+# The setting is particularly useful for tweaking memory settings.
+org.gradle.jvmargs=-Xmx2048m -Dfile.encoding=UTF-8
```
No

yujincheng08 commented on helper/build.gradle.kts:
```diff
@@ -0,0 +1,30 @@
+
+dependencies {
+    compileOnly("androidx.annotation:annotation:1.5.0")
+}
```
tailing newline

KyuubiRan committed:
refactor: new line & remove jvmargs in gradle.properties

yujincheng08 commented on gradle.properties:
```diff
@@ -0,0 +1,20 @@
+# Enables namespacing of each library's R class so that its R class includes only the
+# resources declared in the library itself and none from the library's dependencies,
+# thereby reducing the size of the R class for that library
+android.nonTransitiveRClass=true
```
tailing newline

yujincheng08 replied:
we dont hv resources

yujincheng08 commented on build.gradle.kts:
```diff
@@ -0,0 +1,5 @@
+// Top-level build file where you can add configuration options common to all sub-projects/modules.
+plugins {
+    id("com.android.application") version "8.0.0-alpha11" apply false
```
use stable ver

yujincheng08 replied:
we dont need application? Ignore if you are going to add tests

yujincheng08 commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+#Sun Jan 08 16:38:33 CST 2023
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-bin.zip
```
it's weird that you use alpha AGP but use an old ver of gradle

yujincheng08 commented on helper/.gitignore:
```diff
@@ -0,0 +1 @@
+/build
```
tailing new line

yujincheng08 commented on helper/proguard-rules.pro:
```diff
@@ -0,0 +1,21 @@
+
+# If you keep the line number information, uncomment this to
+# hide the original source file name.
+#-renamesourcefileattribute SourceFile
```
tailing new line

yujincheng08 commented on helper/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
+
+</manifest>
```
tailing new line

yujincheng08 commented on settings.gradle.kts:
```diff
@@ -0,0 +1,16 @@
+        mavenCentral()
+    }
+}
+rootProject.name = "LibXposed"
```
```suggestion
rootProject.name = "libxposed"
```

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ConstructorFinder.java:
```diff
@@ -0,0 +1,333 @@
+    }
+
+    public ConstructorFinder(@NonNull String className, @Nullable ClassLoader classLoader) throws ClassNotFoundException {
+        this(classLoader != null ? classLoader.loadClass(className) : Class.forName(className));
```
Do not initialize the class

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ConstructorFinder.java:
```diff
@@ -0,0 +1,333 @@
+        this.constructorStream = Arrays.stream(constructors);
+    }
+
+    public ConstructorFinder(@NonNull List<Constructor<?>> constructors) {
```
Maybe Collection or Iterable

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ConstructorFinder.java:
```diff
@@ -0,0 +1,333 @@
+     * @param parameterTypeNames parameter type names, use null to skip check some parameters.
+     * @return this
+     */
+    public ConstructorFinder filterByParameterTypeNames(String... parameterTypeNames) {
```
does not support binary name

KyuubiRan committed:
feat: Add BaseFinder.java

KyuubiRan committed:
refactor: cleanup code & change gradle version to 7.6

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,297 @@
+        return (S) this;
+    }
+
+    public S filterByName(@NonNull String name) {
```
constructor has no name

KyuubiRan committed:
refactor: suppress unused in BaseFinder.java

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,297 @@
+     *
+     * @return this
+     */
+    public S filterStatic() {
```
constructor cannot be static

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,297 @@
+     *
+     * @return this
+     */
+    public S filterFinal() {
```
Constructor cannot be final

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,297 @@
+     *
+     * @return this
+     */
+    public S filterSynchronized() {
```
constructor cannot be synchronized

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,297 @@
+     *
+     * @return this
+     */
+    public S filterAbstract() {
```
constructor cannot be abstract

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ConstructorFinder.java:
```diff
@@ -47,8 +42,8 @@ public ConstructorFinder(@NonNull List<Constructor<?>> constructors) {
      */
-    public static ConstructorFinder fromClass(@NonNull Class<?> clazz) {
-        return new ConstructorFinder(clazz);
+    public static <C> ConstructorFinder<C> fromClass(@NonNull Class<C> clazz) {
```
name all `findXXX` to `find`?

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ConstructorFinder.java:
```diff
@@ -5,40 +5,35 @@
-    public ConstructorFinder(@NonNull String className, @Nullable ClassLoader classLoader) throws ClassNotFoundException {
-        this(classLoader != null ? classLoader.loadClass(className) : Class.forName(className));
+    @SuppressWarnings("unchecked")
+    private ConstructorFinder(@NonNull String className, @Nullable ClassLoader classLoader) throws ClassNotFoundException {
```
hand over the dirty jobs to static factories

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/MethodFinder.java:
```diff
@@ -91,37 +85,37 @@ public static MethodFinder fromArray(@NonNull Method[] methods) {
      * @throws IllegalArgumentException If the superclass parameter is not a superclass of the class.
      */
-    public MethodFinder findSuper(@Nullable Class<?> superclass) throws IllegalArgumentException {
+    public MethodFinder findSuper(@Nullable Class<?> untilSuperClass) throws IllegalArgumentException {
```
use a callback to determine the last super class

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ConstructorFinder.java:
```diff
@@ -82,252 +77,24 @@ public static ConstructorFinder fromClassName(@NonNull String className) throws
     @NonNull
-    public <R, A> R collect(@NonNull Collector<? super Constructor<?>, A, R> collector) {
+    @Override
+    public <R, A> R collect(@NonNull Collector<Constructor<C>, A, R> collector) {
```
what if put this to the base class

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,297 @@
+     * @throws NoSuchMethodException if the method/constructor not found
+     */
+    @NonNull
+    public T findFirst() throws NoSuchMethodException {
```
no need `find` since you are not actually finding

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/MethodFinder.java:
```diff
@@ -0,0 +1,160 @@
+     * @return this
+     * @throws IllegalArgumentException If the superclass parameter is not a superclass of the class.
+     */
+    public MethodFinder findSuper(@Nullable Class<?> untilSuperClass) throws IllegalArgumentException {
```
Should move to static factories that initialize string by class

KyuubiRan committed:
refactor: use static methods instead constructors

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/MethodFinder.java:
```diff
@@ -0,0 +1,260 @@
+     * @see #findSuper(Predicate)
+     */
+    public static MethodFinder from(@NonNull Class<?> clazz, @Nullable Predicate<Class<?>> findSuperClassPredicate) {
+        var mf = new MethodFinder(Arrays.stream(clazz.getDeclaredMethods()));
```
why not use raw stream

KyuubiRan committed:
refactor: move collect to base class

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/MethodFinder.java:
```diff
@@ -0,0 +1,244 @@
+     * @return this
+     */
+    public MethodFinder filterNative() {
+        return filter(ModifierHelper::isNative);
```
constructor can be native

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ModifierHelper.java:
```diff
@@ -0,0 +1,154 @@
+ * Helper for quick check the modifier of a member/class.
+ */
+@SuppressWarnings("unused")
+public class ModifierHelper {
```
create a private conatructor for it and throw expection in it to prevent instanization

KyuubiRan committed:
refactor: inline findSuper & private constructor for ModifierHelper

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/MethodFinder.java:
```diff
@@ -0,0 +1,176 @@
+     */
+    public static MethodFinder from(@NonNull Class<?> clazz, @Nullable Predicate<Class<?>> findSuperClassPredicate) {
+        var mf = new MethodFinder(Arrays.stream(clazz.getDeclaredMethods()));
+        if (findSuperClassPredicate != null) {
```
I think this is problematic. Consider the following scenario:

```java
interface A { default char a() { return 'a'; } }
interface B { default char b() { return 'b'; } }

class C implements A, B {}
```

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/MethodFinder.java:
```diff
@@ -0,0 +1,176 @@
+     * @param findSuperClassPredicate find super class predicate(return true = break, false = continue), null if not find
+     * @return MethodFinder
+     */
+    public static MethodFinder from(@NonNull Class<?> clazz, @Nullable Predicate<Class<?>> findSuperClassPredicate) {
```
Use a do while for it

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ModifierHelper.java:
```diff
@@ -0,0 +1,157 @@
+ * Helper for quick check the modifier of a member/class.
+ */
+@SuppressWarnings("unused")
+public class ModifierHelper {
```
Add isPackage

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ModifierHelper.java:
```diff
@@ -0,0 +1,157 @@
+ */
+@SuppressWarnings("unused")
+public class ModifierHelper {
+    private ModifierHelper() {
```
throw an exception inside

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,218 @@
+import java.util.stream.Stream;
+
+@SuppressWarnings({"unchecked", "unused"})
+abstract class BaseFinder<T extends Member, S> {
```
make methods in this class final

yujincheng08 replied:
consider add a filter for getExceptionTypes

KyuubiRan committed:
feat: Add isPackagePrivate

KyuubiRan committed:
refactor: replace findSuperClass by do-while
fix: constructor not found in ConstructorFinder
feat: filter by exception types
feat: filter by package private

KyuubiRan committed:
fix: get default methods in interface.
fix: use Object instead null in find methods of superclass, because some methods in Object are Inaccessible

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/MethodFinder.java:
```diff
@@ -0,0 +1,181 @@
+ * Helper for quick and elegant to find method(s).
+ */
+@SuppressWarnings("unused")
+public final class MethodFinder extends BaseFinder<Method, MethodFinder> {
```
Add filter default

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ModifierHelper.java:
```diff
@@ -0,0 +1,174 @@
+        return !Modifier.isNative(member.getModifiers());
+    }
+
+    public static boolean isInterface(Member member) {
```
Are you sure a member can make this function true?

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ModifierHelper.java:
```diff
@@ -0,0 +1,174 @@
+        return !isPackagePrivate(member);
+    }
+
+    public static boolean isPackagePrivate(Class<?> clazz) {
```
Following AOSP, we should name it `isPackage`

KyuubiRan committed:
feat: filter if method is default

KyuubiRan committed:
refactor: rename isPackagePrivate -> isPackage
fix: remove isInterface(Member)

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,254 @@
+     *
+     * @return this
+     */
+    public final S filterPackagePrivate() {
```
For consistency, name it `filterPackage`

KyuubiRan committed:
refactor: rename filterPackagePrivate -> filterPackage

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,254 @@
+    public final S filterExceptionType(Class<?>... e) {
+        return filter(m -> {
+            var set1 = Set.of(getExceptionTypes(m));
+            var set2 = Set.of(e);
```
Why construct the same set for each loop?

KyuubiRan committed:
fix: construct the same set for each loop

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,254 @@
+     * @param e exception types
+     * @return this
+     */
+    public final S filterExceptionType(Class<?>... e) {
```
Confusing naming. This method is named with `by` but the next method has no `by` in its name.

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/MethodFinder.java:
```diff
@@ -0,0 +1,199 @@
+        do {
+            stream = Stream.concat(stream, Arrays.stream(clazz.getDeclaredMethods()));
+            stream = Stream.concat(stream, Arrays.stream(clazz.getInterfaces())
+                    .flatMap(c -> Arrays.stream(c.getDeclaredMethods()).filter(Method::isDefault)));
```
Inconsistency. For classes, you don't filter abstract methods, but for interfaces, you do.

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,254 @@
+import java.util.stream.Stream;
+
+@SuppressWarnings({"unchecked", "unused"})
+abstract class BaseFinder<T extends Member, S> {
```
Consider adding varargs filter.

yujincheng08 replied:
Consider adding varargs filter

KyuubiRan committed:
refactor: filterExceptionType -> filterByExceptionType

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,254 @@
+import java.util.stream.Stream;
+
+@SuppressWarnings({"unchecked", "unused"})
+abstract class BaseFinder<T extends Member, S> {
```
For extendibility, consider adding `includes` and `excludes` filter for parameter types and exception types.

yujincheng08 replied:
This one is not resolved.

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,254 @@
+     * @param parameterCount parameter count
+     * @return this
+     */
+    public final S filterByParameterCount(int parameterCount) {
```
What if I need to filter methods with parameter counts in range? consider filter by a predicate.

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/BaseFinder.java:
```diff
@@ -0,0 +1,254 @@
+     * @param parameterTypes parameter types, use null to skip check some parameters.
+     * @return this
+     */
+    public final S filterByParameterTypes(Class<?>... parameterTypes) {
```
I still think this is not extendable. What if I want a method with two ints in its parameters? Indices-specific filtering is not extendable. How about filtering by predicate, but we pre-define some useful predicates like full matcher, include maters, and exclude matcher? Same for exception types.

KyuubiRan replied:
I think this is a reasonable demand, parameter types must full-match(or null if skip).
```java
var method = MethodFinder.from(String.class, null)
            .filterByParameterTypes(char[].class, null, int.class)
            .first();
System.out.println(method.getName());

var method2 = MethodFinder.from(String.class, null)
            .filterByParameterTypes(char[].class, int.class, int.class)
            .first();
System.out.println(method2.getName());
```
output:
```
valueOf
valueOf
````
it‘s easier to use than predicate

KyuubiRan replied:
if you want use more condition, why not use `BaseFinder#filter(Predicate<T>)` ?

yujincheng08 commented on gradlew:
```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env sh
```
This file should be executable!

yujincheng08 commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+#Sun Jan 08 16:38:33 CST 2023
```
No timestamp comment here.

yujincheng08 commented on helper/build.gradle.kts:
```diff
@@ -0,0 +1,30 @@
+
+    buildTypes {
+        named("release") {
+            isMinifyEnabled = true
```
This is a library. If you enable minify without correct proguard rules (currently empty), how can users use your library?

KyuubiRan committed:
feat: Add isVarargs & filterVarargs
refactor: reformat code & rename filterExceptionType -> filterByExceptionType
fix: filter abstract interface methods

KyuubiRan committed:
feat: Add filterByParameterTypes(Predicate)

yujincheng08 commented on helper/src/main/java/io/github/libxposed/helpers/ModifierHelper.java:
```diff
@@ -163,4 +163,13 @@ public static boolean isPackage(Class<?> clazz) {
         return !isPackage(clazz);
     }
+
+    public static boolean isVarargs(Member member) {
```
Maybe also add corresponding functions in ModifierHelper

yujincheng08 commented on helper/build.gradle.kts:
```diff
@@ -15,7 +15,7 @@ android {
     buildTypes {
         named("release") {
-            isMinifyEnabled = true
+            isMinifyEnabled = false
```
We can still minify the library to reduce size. What we need to fix is to write the correct proguard.

yujincheng08 commented on gradlew:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env sh
+#!/bin/sh
```
why such a change?
